cardano-db-sync/src/Cardano/DbSync/Era/Universal/Insert/Other.hs

```diff
@@ -125,53 +132,71 @@ insertDatum tracer cache txId txd = do
   (MonadBaseControl IO m, MonadIO m) =>
-  Trace IO Text ->
+  SyncEnv ->
   Cache ->
   DB.TxId ->
   Map Word64 DB.RedeemerId ->
   Generic.TxWithdrawal ->
   ExceptT SyncNodeError (ReaderT SqlBackend m) ()
-insertWithdrawals _tracer cache txId redeemers txWdrl = do
-  addrId <-
-    lift $ queryOrInsertRewardAccount cache CacheNew $ Generic.txwRewardAccount txWdrl
-  void . lift . DB.insertWithdrawal $
-    DB.Withdrawal
-      { DB.withdrawalAddrId = addrId
-      , DB.withdrawalTxId = txId
-      , DB.withdrawalAmount = Generic.coinToDbLovelace $ Generic.txwAmount txWdrl
-      , DB.withdrawalRedeemerId = mlookup (Generic.txwRedeemerIndex txWdrl) redeemers
-      }
+insertWithdrawals syncEnv cache txId redeemers txWdrl = do
+  -- check if shelley stake address is in the whitelist
+  when (shelleyStakeAddrWhitelistCheck syncEnv $ Generic.txwRewardAccount txWdrl) $ do
+    addrId <-
+      lift $ queryOrInsertRewardAccount syncEnv cache CacheNew $ Generic.txwRewardAccount txWdrl
+    void
+      . lift
+      . DB.insertWithdrawal
+      $ DB.Withdrawal
+        { DB.withdrawalAddrId = addrId
+        , DB.withdrawalTxId = txId
+        , DB.withdrawalAmount = Generic.coinToDbLovelace $ Generic.txwAmount txWdrl
+        , DB.withdrawalRedeemerId = mlookup (Generic.txwRedeemerIndex txWdrl) redeemers
+        }
 
 -- | Insert a stake address if it is not already in the `stake_address` table. Regardless of
 -- whether it is newly inserted or it is already there, we retrun the `StakeAddressId`.
 insertStakeAddressRefIfMissing ::
   (MonadBaseControl IO m, MonadIO m) =>
-  Trace IO Text ->
+  SyncEnv ->
   Cache ->
   Ledger.Addr StandardCrypto ->
   ReaderT SqlBackend m (Maybe DB.StakeAddressId)
-insertStakeAddressRefIfMissing _trce cache addr =
+insertStakeAddressRefIfMissing syncEnv cache addr = do
   case addr of
     Ledger.AddrBootstrap {} -> pure Nothing
     Ledger.Addr nw _pcred sref ->
       case sref of
         Ledger.StakeRefBase cred -> do
-          Just <$> queryOrInsertStakeAddress cache DontCacheNew nw cred
+          -- Check if the stake address is in the shelley whitelist
+          if shelleyStakeAddrWhitelistCheck syncEnv $ Ledger.RewardAccount nw cred
+            then do
+              Just <$> queryOrInsertStakeAddress syncEnv cache DontCacheNew nw cred
+            else pure Nothing
         Ledger.StakeRefPtr ptr -> do
           queryStakeRefPtr ptr
         Ledger.StakeRefNull -> pure Nothing
 
 insertMultiAsset ::
   (MonadBaseControl IO m, MonadIO m) =>
   Cache ->
+  Maybe (NonEmpty ShortByteString) ->
   PolicyID StandardCrypto ->
   AssetName ->
-  ReaderT SqlBackend m DB.MultiAssetId
-insertMultiAsset cache policy aName = do
+  ReaderT SqlBackend m (Maybe DB.MultiAssetId)
+insertMultiAsset cache mWhitelist policy aName = do
   mId <- queryMAWithCache cache policy aName
   case mId of
-    Right maId -> pure maId
+    Right maId -> pure $ Just maId
     Left (policyBs, assetNameBs) ->
+      case mWhitelist of
```

Comment:
Do we even need this check? It seems to be done already at `isMAPoliciesInWhitelist` (but possibly it needs some extension there to cover minting, see the comment there). Quite possible we don't reach this place if the multiasset is not whitelisted already.

Reply:
> see the comment there
The other comments don't seem to have registered as I can't see them

Reply:
This should be it https://github.com/IntersectMBO/cardano-db-sync/pull/1644/files/c752e61eb0f38392c4c2d823a395fc2b6178541f#r1599641782

Reply:
inside of `insertTxOut` I do:
```
case ioMultiAssets iopts of
        MultiAssetDisable -> pure $ Just (eutxo, mempty)
        MultiAssetEnable -> do
          !maTxOuts <- insertMaTxOuts cache Nothing maMap
          pure $ Just (eutxo, maTxOuts)
        MultiAssetPolicies whitelist -> do
          !maTxOuts <- insertMaTxOuts cache (Just whitelist) maMap
          pure $ Just (eutxo, maTxOuts)
```
This is just to ascertain if a whitelist exists and if so pass it along. The reason I only pass the whitelist is because not until this code:
```
mId <- queryMAWithCache cache policy aName
case mId of
    Right maId -> pure $ Just maId
    Left (policyBs, assetNameBs) ->
      case mWhitelist of
```
do I get access to `policyBs` which is what is used to check against whitelist.

Prior I used to have the whitelist check in `queryMAWithCache` but the problem there is that not all uses of queryMAWithCache require a whitelist check and that caused all sorts of errors 😞

Reply:
I think it's similar to https://github.com/IntersectMBO/cardano-db-sync/pull/1644#discussion_r1599790322: it's just different behaviours that it can have. And thinking again the behaviour defined is not wrong.